cxx-squid/src/main/java/org/sonar/cxx/parser/CxxGrammarImpl.java

```diff
@@ -502,15 +502,12 @@ private static void expressions(LexerfulGrammarBuilder b) {
     b.rule(nestedNameSpecifier).is(
       b.firstOf(
         "::",
-        b.sequence(typeName, "::"), // C++
+        b.sequence(className, "::"), // C++
```

Comment:
Unfortunately `nestedNameSpecifier` was not expressed in terms of `className`s. 

* `typeName` is theoretically a super-set w.r.t. the `className`but practically it's completely interchangeable with `className`. See below




```JAVA
    b.rule(typeName).is(
      b.firstOf(
        className,       // == b.firstOf( simpleTemplateId, IDENTIFIER )
        enumName,        // == IDENTIFIER
        typedefName,     // == IDENTIFIER
        simpleTemplateId // 
      )
    );
```
but the real problem lies below...